File: Algorithm_study_01/10_20200225과제물/등굣길.java
class Solution {
    public int solution(int m, int n, int[][] puddles) {
        int answer;
        int dp[][]= new int[101][101];
        int maps[][]=new int[101][101];
        
        //물웅덩이 => -1로 한다.
        //(px,py)
        for(int y=0; y<puddles.length; y++)
            maps[puddles[y][1]][puddles[y][0]]=-1;
        
        dp[1][0]=1;
        for(int y=1; y<=n; y++){
            for(int x=1; x<=m; x++){
                if (maps[y][x]==-1){
                    dp[y][x]=0;
                }
                    
                else{
                    dp[y][x]=(dp[y-1][x]+dp[y][x-1])%1000000007;
                }
            }
        }
        return dp[n][m];
    }
}

File: Algorithm_study_01/10_20200225과제물/등굣길.py
def isRange(y,x, n, m):
    if(y>=0 and y<n) and (x>=0 and x<m):
        return True
    return False

def solution(m, n, puddles):
    maps=[[0 for x in range(m)] for y in range(n)]#지도초기화
    dp=[[0 for x in range(m)] for y in range(n)] #경로 초기화
    
    for px, py in puddles:#물웅덩이 위치의 maps값 -1
        maps[py-1][px-1]=-1
    
    
    dy=(-1,0) #위, 왼쪽
    dx=(0,-1)
    dp[0][0]=1
    for y in range(n):
        for x in range(m):
            if maps[y][x]!=-1:
                for i in range(2):
                    prevy=dy[i]+y
                    prevx=dx[i]+x
                    if isRange(prevy,prevx,n,m):
                        if maps[prevy][prevx]!=-1:
                            dp[y][x]+=dp[prevy][prevx]
                dp[y][x]%=1000000007
                
    answer=dp[n-1][m-1]
    return answer
